File: backend/schemas/product_truth_job_confirm.py
"""Schemas for ConfirmJobProductTruth — job-level Product Truth revision."""

from __future__ import annotations

from typing import Any, Literal, Optional

from pydantic import BaseModel, Field


JobConfirmationState = Literal["unconfirmed", "confirmed", "stale_after_edit"]


class ConfirmJobProductTruthRequest(BaseModel):
    expected_revision: int = Field(ge=0, description="0 for first confirm; else current revision")
    expected_draft_hash: Optional[str] = Field(
        default=None,
        description="Optional optimistic concurrency guard against draft payload hash",
    )
    expected_content_hash: Optional[str] = Field(
        default=None,
        description="Optional guard when re-confirming / correcting an existing revision",
    )
    root_template_code: Optional[str] = None
    correction_reason: Optional[str] = None


class JobRevisionMetadataView(BaseModel):
    revision: int
    content_hash: str
    confirmation_state: str
    confirmed_at: Optional[str] = None
    confirmed_by: Optional[str] = None
    expected_draft_hash: Optional[str] = None
    root_template_code: Optional[str] = None
    root_template_version: Optional[str] = None
    contract_version: str = "job_revision_v1"
    source: str = "confirm_job_product_truth"
    provenance: dict[str, Any] = Field(default_factory=dict)


class ConfirmJobProductTruthResponse(BaseModel):
    workspace_id: str
    workspace_code: Optional[str] = None
    write_performed: bool
    idempotent_noop: bool
    product_truth_path: str = "payload_json.product_truth.confirmed_snapshot_v1"
    metadata: JobRevisionMetadataView
    pinned_bag_keys: list[str] = Field(default_factory=list)
    draft_hash: str
    previous_revision: Optional[int] = None
    audit_entry: Optional[dict[str, Any]] = None


class JobProductTruthStatusResponse(BaseModel):
    workspace_id: str
    has_job_revision: bool
    metadata: Optional[JobRevisionMetadataView] = None
    draft_hash: str
    is_stale: bool = False
    commercial_freeze_allowed: bool = False

File: frontend/src/lib/intakeV6/productTruthJobConfirm.ts
/**
 * ConfirmJobProductTruth client — job-level Product Truth revision.
 * Catalog is never written. No ProductInstance tables.
 */

import { getAPIBaseURL } from "@/lib/config";

export type JobConfirmationState = "unconfirmed" | "confirmed" | "stale_after_edit" | string;

export interface JobRevisionMetadata {
  revision: number;
  content_hash: string;
  confirmation_state: JobConfirmationState;
  confirmed_at?: string | null;
  confirmed_by?: string | null;
  expected_draft_hash?: string | null;
  root_template_code?: string | null;
  root_template_version?: string | null;
  contract_version?: string;
  source?: string;
  provenance?: Record<string, unknown>;
}

export interface ConfirmJobProductTruthResponse {
  workspace_id: string;
  workspace_code?: string | null;
  write_performed: boolean;
  idempotent_noop: boolean;
  metadata: JobRevisionMetadata;
  pinned_bag_keys: string[];
  draft_hash: string;
  previous_revision?: number | null;
}

export interface JobProductTruthStatusResponse {
  workspace_id: string;
  has_job_revision: boolean;
  metadata?: JobRevisionMetadata | null;
  draft_hash: string;
  is_stale: boolean;
  commercial_freeze_allowed: boolean;
}

function apiBase(): string {
  return `${getAPIBaseURL()}/api/v1/intake-v6`;
}

async function requestJson<T>(url: string, init?: RequestInit): Promise<T> {
  const res = await fetch(url, {
    credentials: "include",
    ...init,
  });
  if (!res.ok) {
    let detail: unknown = null;
    try {
      detail = await res.json();
    } catch {
      detail = await res.text();
    }
    const err = new Error(`HTTP ${res.status}`) as Error & {
      status: number;
      detail: unknown;
    };
    err.status = res.status;
    err.detail = detail;
    throw err;
  }
  return (await res.json()) as T;
}

export async function getJobProductTruthStatus(
  workspaceId: string,
): Promise<JobProductTruthStatusResponse> {
  return requestJson(
    `${apiBase()}/workspaces/${encodeURIComponent(workspaceId)}/product-truth/job-status`,
  );
}

export async function confirmJobProductTruth(
  workspaceId: string,
  body: {
    expected_revision: number;
    expected_draft_hash?: string | null;
    expected_content_hash?: string | null;
    root_template_code?: string | null;
    correction_reason?: string | null;
  },
): Promise<ConfirmJobProductTruthResponse> {
  return requestJson(
    `${apiBase()}/workspaces/${encodeURIComponent(workspaceId)}/product-truth/confirm-job`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    },
  );
}
